src/env.cc

```diff
@@ -300,6 +310,29 @@ void Environment::PrintSyncTrace() const {
   fflush(stderr);
 }
 
+void Environment::RunCleanup() {
+  CleanupHandles();
+
+  while (!cleanup_hooks_.empty()) {
+    // Copy into a vector, since we can't sort an unordered_set in-place.
+    std::vector<CleanupHookCallback> callbacks(
+        cleanup_hooks_.begin(), cleanup_hooks_.end());
+    cleanup_hooks_.clear();
+
+    std::sort(callbacks.begin(), callbacks.end(),
+              [](const CleanupHookCallback& a, const CleanupHookCallback& b) {
+      // Sort in descending order so that the most recently inserted callbacks
+      // are run first.
+      return a.insertion_order_counter_ > b.insertion_order_counter_;
+    });
+
+    for (const CleanupHookCallback& cb : callbacks) {
+      cb.fn_(cb.arg_);
+      CleanupHandles();
```

Comment:
@addaleax One question here:
`CleanupHandles()` may be executed multiple times here plus one time (the beginning of `RunCleanup()`). Inside the `CleanupHandles()` it iterates through several queues. Some of them call clear() after the for loop. Some of them not. Is it necessary to go through `req_wrap_queue_` and `handle_wrap_queue_` every time?

Reply:
@yhwang I’m not sure, I can try to figure out whether this was necessary

I guess the `CleanupHandles()` call could be moved after this loop…